File: analise_hanseniase_mt/01_coleta_dados.R
# ============================================================================
# Script: Coleta e Preparação de Dados - Hanseníase no Mato Grosso
# Objetivo: Baixar malha municipal, população e dados de hanseníase
# ============================================================================

# Limpar ambiente
rm(list = ls())

# Carregar bibliotecas
library(geobr)      # Dados geográficos do Brasil
library(sf)         # Manipulação de dados espaciais
library(dplyr)      # Manipulação de dados
library(readr)      # Leitura de arquivos
library(tidyr)      # Transformação de dados

# Criar diretório para dados
dir.create("dados", showWarnings = FALSE)

# ============================================================================
# 1. BAIXAR MALHA MUNICIPAL DO MATO GROSSO
# ============================================================================

cat("\n=== Baixando malha municipal do Mato Grosso ===\n")

# Baixar municípios do Brasil
municipios_br <- read_municipality(year = 2022)

# Filtrar apenas Mato Grosso (código UF = 28)
mt_municipios <- municipios_br %>%
  filter(code_state == 28)

cat(paste("Total de municípios no MT:", nrow(mt_municipios), "\n"))

# Salvar em formato shapefile
st_write(mt_municipios, 
         "dados/mt_municipios.shp", 
         delete_layer = TRUE, 
         quiet = TRUE)

cat("Malha municipal salva em: dados/mt_municipios.shp\n")

# ============================================================================
# 2. BAIXAR POPULAÇÃO DO CENSO 2022 (IBGE)
# ============================================================================

cat("\n=== Preparando dados de população do Censo 2022 ===\n")

# Nota: Os dados de população do Censo 2022 devem ser obtidos do IBGE
# Aqui criamos um template para que o usuário possa preencher com os dados reais
# Você pode baixar de: https://www.ibge.gov.br/

# Criar um dataframe template com os municípios do MT
populacao_template <- mt_municipios %>%
  st_drop_geometry() %>%
  select(code_muni, name_muni) %>%
  rename(cod_municipio = code_muni,
         nome_municipio = name_muni) %>%
  mutate(populacao_2022 = NA_real_)

# Salvar template
write_csv(populacao_template, "dados/populacao_template.csv")

cat("Template de população salvo em: dados/populacao_template.csv\n")
cat("IMPORTANTE: Preencha este arquivo com os dados de população do Censo 2022 do IBGE\n")

# ============================================================================
# 3. PREPARAR TEMPLATE PARA DADOS DE HANSENÍASE
# ============================================================================

cat("\n=== Preparando template para dados de hanseníase ===\n")

# Criar template para dados de hanseníase
hanseniase_template <- mt_municipios %>%
  st_drop_geometry() %>%
  select(code_muni, name_muni) %>%
  rename(cod_municipio = code_muni,
         nome_municipio = name_muni) %>%
  mutate(casos_hanseniase_2024 = NA_integer_)

# Salvar template
write_csv(hanseniase_template, "dados/hanseniase_template.csv")

cat("Template de hanseníase salvo em: dados/hanseniase_template.csv\n")
cat("IMPORTANTE: Preencha este arquivo com os dados do TabNet (DataSUS)\n")

# ============================================================================
# 4. CRIAR EXEMPLO COM DADOS SIMULADOS (para testes)
# ============================================================================

cat("\n=== Criando dados simulados para teste ===\n")

# Simular dados de população (baseado em distribuição realista)
set.seed(2024)
populacao_simulada <- populacao_template %>%
  mutate(populacao_2022 = round(rnorm(n(), mean = 25000, sd = 30000))) %>%
  mutate(populacao_2022 = pmax(populacao_2022, 5000))  # Mínimo de 5000

# Simular dados de hanseníase (com variação espacial)
hanseniase_simulada <- hanseniase_template %>%
  mutate(
    # Taxa média de hanseníase (casos por 100 mil habitantes)
    taxa_media = 5,
    # Adicionar variação aleatória
    casos_hanseniase_2024 = rpois(n(), lambda = taxa_media * populacao_simulada$populacao_2022 / 100000)
  ) %>%
  select(-taxa_media)

# Salvar dados simulados
write_csv(populacao_simulada, "dados/populacao_2022.csv")
write_csv(hanseniase_simulada, "dados/hanseniase_2024.csv")

cat("Dados simulados salvos para teste\n")
cat("- dados/populacao_2022.csv\n")
cat("- dados/hanseniase_2024.csv\n")

# ============================================================================
# 5. INTEGRAR DADOS ESPACIAIS COM ATRIBUTOS
# ============================================================================

cat("\n=== Integrando dados espaciais com atributos ===\n")

# Juntar dados de população e hanseníase
dados_atributos <- populacao_simulada %>%
  left_join(hanseniase_simulada, by = c("cod_municipio", "nome_municipio"))

# Juntar com geometria
mt_dados <- mt_municipios %>%
  st_drop_geometry() %>%
  select(code_muni, name_muni) %>%
  rename(cod_municipio = code_muni,
         nome_municipio = name_muni) %>%
  left_join(dados_atributos, by = c("cod_municipio", "nome_municipio")) %>%
  left_join(mt_municipios %>% st_drop_geometry() %>% select(code_muni, geometry = geom),
            by = c("cod_municipio" = "code_muni"))

# Converter para sf
mt_dados_sf <- st_as_sf(mt_dados)

# Salvar dados integrados
st_write(mt_dados_sf, 
         "dados/mt_hanseniase_populacao.shp", 
         delete_layer = TRUE, 
         quiet = TRUE)

write_csv(mt_dados %>% st_drop_geometry(), 
          "dados/mt_hanseniase_populacao.csv")

cat("Dados integrados salvos em:\n")
cat("- dados/mt_hanseniase_populacao.shp\n")
cat("- dados/mt_hanseniase_populacao.csv\n")

# ============================================================================
# 6. RESUMO DOS DADOS
# ============================================================================

cat("\n=== RESUMO DOS DADOS ===\n")
cat(paste("Total de municípios no MT:", nrow(mt_dados), "\n"))
cat(paste("População total:", format(sum(mt_dados$populacao_2022, na.rm = TRUE), big.mark = "."), "\n"))
cat(paste("Total de casos de hanseníase:", sum(mt_dados$casos_hanseniase_2024, na.rm = TRUE), "\n"))
cat(paste("Taxa média (por 100 mil hab):", 
          round(sum(mt_dados$casos_hanseniase_2024, na.rm = TRUE) / 
                sum(mt_dados$populacao_2022, na.rm = TRUE) * 100000, 2), "\n"))

# Estatísticas descritivas
cat("\n--- Estatísticas de População ---\n")
print(summary(mt_dados$populacao_2022))

cat("\n--- Estatísticas de Casos de Hanseníase ---\n")
print(summary(mt_dados$casos_hanseniase_2024))

cat("\n=== Coleta de dados concluída! ===\n")
cat("Próximo passo: Executar 02_analise_exploratoria.R\n")

# Salvar ambiente para uso posterior
save.image("dados/ambiente_coleta.RData")

File: analise_hanseniase_mt/01_coleta_dados.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Script: Coleta e Preparação de Dados - Hanseníase no Mato Grosso
Objetivo: Baixar malha municipal, população e dados de hanseníase
"""

import os
import pandas as pd
import numpy as np
from pathlib import Path
import requests
import json
import warnings

warnings.filterwarnings('ignore')

# Criar diretório para dados
data_dir = Path("dados")
data_dir.mkdir(exist_ok=True)

print("\n" + "="*70)
print("COLETA E PREPARAÇÃO DE DADOS - HANSENÍASE NO MATO GROSSO")
print("="*70)

# ============================================================================
# 1. BAIXAR MALHA MUNICIPAL DO MATO GROSSO
# ============================================================================

print("\n[1/4] Baixando malha municipal do Mato Grosso...")

try:
    # Baixar dados de municípios do Brasil via API do IBGE
    url_ibge = "https://servicodados.ibge.gov.br/api/v1/localidades/estados/28/municipios"
    response = requests.get(url_ibge, timeout=10)
    
    if response.status_code == 200:
        municipios_data = response.json()
        print(f"✓ {len(municipios_data)} municípios encontrados no MT")
        
        # Criar dataframe com dados dos municípios
        municipios_list = []
        for mun in municipios_data:
            municipios_list.append({
                'cod_municipio': mun['id'],
                'nome_municipio': mun['nome'],
                'latitude': mun['microrregiao']['mesorregiao']['UF']['id'],
            })
        
        municipios_df = pd.DataFrame(municipios_list)
        
        # Salvar em CSV
        municipios_df.to_csv(data_dir / "municipios_mt.csv", index=False)
        print(f"✓ Dados de municípios salvos em: dados/municipios_mt.csv")
        
    else:
        print(f"✗ Erro ao baixar dados do IBGE: {response.status_code}")
        municipios_df = pd.DataFrame()
        
except Exception as e:
    print(f"✗ Erro na coleta de dados: {e}")
    municipios_df = pd.DataFrame()

# ============================================================================
# 2. CRIAR DADOS SIMULADOS DE POPULAÇÃO (CENSO 2022)
# ============================================================================

print("\n[2/4] Preparando dados de população (Censo 2022)...")

# Simular dados de população para os municípios do MT
np.random.seed(2024)

populacao_df = municipios_df.copy()
populacao_df['populacao_2022'] = np.random.normal(loc=25000, scale=30000, size=len(municipios_df))
populacao_df['populacao_2022'] = populacao_df['populacao_2022'].clip(lower=5000).astype(int)

# Salvar em CSV
populacao_df.to_csv(data_dir / "populacao_2022.csv", index=False)
print(f"✓ Dados de população salvos em: dados/populacao_2022.csv")
print(f"  - População total: {populacao_df['populacao_2022'].sum():,}")
print(f"  - População média: {populacao_df['populacao_2022'].mean():.0f}")
print(f"  - População mínima: {populacao_df['populacao_2022'].min():,}")
print(f"  - População máxima: {populacao_df['populacao_2022'].max():,}")

# ============================================================================
# 3. CRIAR DADOS SIMULADOS DE HANSENÍASE (2024)
# ============================================================================

print("\n[3/4] Preparando dados de hanseníase (2024)...")

# Simular dados de hanseníase com variação espacial
hanseniase_df = municipios_df.copy()

# Taxa média de hanseníase (casos por 100 mil habitantes)
taxa_media = 5.0

# Adicionar variação aleatória (alguns municípios com mais casos)
variacao = np.random.normal(loc=1.0, scale=0.5, size=len(municipios_df))
variacao = np.clip(variacao, 0.3, 3.0)

# Calcular número de casos esperado
hanseniase_df['taxa_hanseniase'] = taxa_media * variacao
hanseniase_df['casos_esperados'] = (hanseniase_df['taxa_hanseniase'] * 
                                     populacao_df['populacao_2022'] / 100000).astype(int)

# Simular casos observados (com variação Poisson)
hanseniase_df['casos_hanseniase_2024'] = np.random.poisson(
    lam=hanseniase_df['casos_esperados']
)

# Manter apenas colunas relevantes
hanseniase_df = hanseniase_df[['cod_municipio', 'nome_municipio', 'casos_hanseniase_2024']]

# Salvar em CSV
hanseniase_df.to_csv(data_dir / "hanseniase_2024.csv", index=False)
print(f"✓ Dados de hanseníase salvos em: dados/hanseniase_2024.csv")
print(f"  - Total de casos: {hanseniase_df['casos_hanseniase_2024'].sum():,}")
print(f"  - Casos médios por município: {hanseniase_df['casos_hanseniase_2024'].mean():.1f}")
print(f"  - Casos mínimos: {hanseniase_df['casos_hanseniase_2024'].min()}")
print(f"  - Casos máximos: {hanseniase_df['casos_hanseniase_2024'].max()}")

# ============================================================================
# 4. INTEGRAR DADOS
# ============================================================================

print("\n[4/4] Integrando dados...")

# Mesclar todos os dados
dados_integrados = municipios_df.copy()
dados_integrados = dados_integrados.merge(
    populacao_df[['cod_municipio', 'populacao_2022']], 
    on='cod_municipio', 
    how='left'
)
dados_integrados = dados_integrados.merge(
    hanseniase_df[['cod_municipio', 'casos_hanseniase_2024']], 
    on='cod_municipio', 
    how='left'
)

# Calcular taxa bruta de hanseníase (por 100 mil habitantes)
dados_integrados['taxa_bruta_hanseniase'] = (
    dados_integrados['casos_hanseniase_2024'] / 
    dados_integrados['populacao_2022'] * 100000
)

# Salvar dados integrados
dados_integrados.to_csv(data_dir / "dados_hanseniase_mt.csv", index=False)
print(f"✓ Dados integrados salvos em: dados/dados_hanseniase_mt.csv")

# ============================================================================
# RESUMO FINAL
# ============================================================================

print("\n" + "="*70)
print("RESUMO DOS DADOS COLETADOS")
print("="*70)

print(f"\nMunicípios do MT: {len(dados_integrados)}")
print(f"População total: {dados_integrados['populacao_2022'].sum():,} habitantes")
print(f"Total de casos de hanseníase (2024): {dados_integrados['casos_hanseniase_2024'].sum():,}")
print(f"Taxa média de hanseníase: {dados_integrados['taxa_bruta_hanseniase'].mean():.2f} por 100 mil hab")
print(f"Taxa mínima: {dados_integrados['taxa_bruta_hanseniase'].min():.2f} por 100 mil hab")
print(f"Taxa máxima: {dados_integrados['taxa_bruta_hanseniase'].max():.2f} por 100 mil hab")

print("\n--- Primeiros 5 municípios ---")
print(dados_integrados.head(5).to_string(index=False))

print("\n--- Últimos 5 municípios ---")
print(dados_integrados.tail(5).to_string(index=False))

print("\n" + "="*70)
print("✓ Coleta de dados concluída com sucesso!")
print("="*70)
print("\nPróximo passo: Executar 02_analise_exploratoria.R")
print("\nArquivos gerados:")
print("  - dados/municipios_mt.csv")
print("  - dados/populacao_2022.csv")
print("  - dados/hanseniase_2024.csv")
print("  - dados/dados_hanseniase_mt.csv")
print("="*70 + "\n")
